File: functions/SSA_my.py
"""
SSA анализ - колебания
"""
import numpy as np
from .Huang_f import Huang_f
from utils import setup_figure, finalize_figure, iter_windows
from config import config


def SSA_my(Y, start_offset, window_shift, window_width, num_windows, fig=None):
    Y = np.array(Y).flatten()
    s = config.plots.style
    ssa_threshold = config.constants.ssa_threshold

    s0 = []
    y_h = []

    for y in iter_windows(Y, start_offset, window_shift, window_width, num_windows):
        s0.append(Huang_f(y))
        y_d = np.diff(y, n=1)
        y_h.append(np.max(y_d) - np.min(y_d) if len(y_d) > 0 else 0)

    s0 = np.array(s0)
    y_h = np.array(y_h)

    Js = np.where(s0 < ssa_threshold)[0]
    ss = np.zeros_like(s0)
    if len(Js) > 0 and len(y_h) > 0:
        ss[Js] = np.mean(y_h)

    fig, [ax1, ax2], standalone = setup_figure(fig, 2, 1, figsize=(12, 8))

    ax1.plot(y_h, linewidth=s.thick_linewidth, label='Размах разностей')
    ax1.plot(ss, 'r', linewidth=s.thick_linewidth, label='Участки с малыми разностями')
    ax1.grid(True)
    ax1.legend()

    ax2.plot(s0, linewidth=s.thick_linewidth)
    ax2.grid(True)
    ax2.set_title('Разности логарифмов сингулярных чисел',
                  fontsize=s.title_fontsize)
    ax2.set_xlabel('Время (с)', fontsize=s.label_fontsize)

    finalize_figure(fig, standalone)

File: matlab/SSA_my.m
function SSA_my(Y,K,d,L,R);

for k=1:R;
    y=Y(K+(k-1)*d+1:K+(k-1)*d+L,1);   %y - столбец
    s0(k)=Huang_f(y);        %минимальная разность сингулярных чисел
    y_d=diff(y,1); 
   %y_d=diff(y,2);
    y_h(k)=max(y_d)-min(y_d);    
end;
Js=find(s0<0.005);
ss=zeros(size(s0)); ss(Js)=ones(size(Js))*mean(y_h); 

hF6=figure; 
set(hF6,'Units','Normalized','Pos',[0.2 0.175 0.78 0.6],...
    'Color',[0.8 0.8 0.8]);
subplot(2,1,1);
plot(y_h,'LineWidth',4); grid; %axis([0 R 0 1]);

hold on; plot(ss,'r','LineWidth',4);

subplot(2,1,2);
plot(s0,'LineWidth',4); grid; %axis([0 R 0 1]);
title('Разности логарифмов сингулярных чисел','FontSize',14);
xlabel('Время (с)','FontSize',14); 